homeassistant/components/samsungtv/config_flow.py

```diff
@@ -96,81 +153,89 @@ async def async_step_import(self, user_input=None):
-        self._manufacturer = discovery_info.get(ATTR_UPNP_MANUFACTURER)
-        self._model = discovery_info.get(ATTR_UPNP_MODEL_NAME)
-        self._name = f"Samsung {self._model}"
-        self._id = discovery_info.get(ATTR_UPNP_UDN)
-        self._title = self._model
+    async def async_step_ssdp(self, user_input=None):
+        """Handle a flow initialized by ssdp discovery."""
+        self._host = urlparse(user_input[ATTR_SSDP_LOCATION]).hostname
+        self._id = user_input.get(ATTR_UPNP_UDN)
 
         # probably access denied
         if self._id is None:
             return self.async_abort(reason=RESULT_AUTH_MISSING)
         if self._id.startswith("uuid:"):
             self._id = self._id[5:]
 
-        await self.async_set_unique_id(ip_address)
-        self._abort_if_unique_id_configured(
-            {
-                CONF_ID: self._id,
-                CONF_MANUFACTURER: self._manufacturer,
-                CONF_MODEL: self._model,
-            }
-        )
+        await self.async_set_unique_id(self._id)
+        await self._abort_if_already_configured()
+
+        await self._get_and_check_device_info()
+
+        self._manufacturer = user_input.get(ATTR_UPNP_MANUFACTURER)
+        if not self._model:
+            self._model = user_input.get(ATTR_UPNP_MODEL_NAME)
+
+        if self._model.startswith(MODEL_SOUNDBAR):
+            return self.async_abort(reason=RESULT_NOT_SUPPORTED)
+
+        self._name = f"{self._manufacturer} {self._model}"
+        self._title = self._model
+
+        self.context["title_placeholders"] = {"model": self._model}
+        return await self.async_step_confirm()
+
+    async def async_step_zeroconf(self, user_input=None):
+        """Handle a flow initialized by zeroconf discovery."""
+        self._host = user_input[CONF_HOST]
+        self._id = user_input[ATTR_PROPERTIES].get("serialNumber")
+
+        if self._id:
+            await self.async_set_unique_id(self._id)
+            await self._abort_if_already_configured()
+
+        await self._get_and_check_device_info()
+
+        self._mac = user_input[ATTR_PROPERTIES].get("deviceid")
+        self._manufacturer = user_input[ATTR_PROPERTIES].get("manufacturer")
+        if not self._model:
+            self._model = user_input[ATTR_PROPERTIES].get("model")
+        self._name = f"{self._manufacturer} {self._model}"
+        self._title = self._model
 
         self.context["title_placeholders"] = {"model": self._model}
         return await self.async_step_confirm()
 
     async def async_step_confirm(self, user_input=None):
         """Handle user-confirmation of discovered node."""
         if user_input is not None:
-            result = await self.hass.async_add_executor_job(self._try_connect)
+            await self.hass.async_add_executor_job(self._try_connect)
 
-            if result != RESULT_SUCCESS:
-                return self.async_abort(reason=result)
             return self._get_entry()
 
```

Comment:
Try to connect here and abort if you can't since showing the form will only result in failure.

Then you can also add

`self._set_confirm_only()`